pkg/relayer/relayminer.go

```diff
@@ -0,0 +1,64 @@
+package relayer
+
+import (
+	"context"
+	"log"
+
+	"cosmossdk.io/depinject"
+)
+
+// relayMiner is the main struct that encapsulates the relayer's responsibilities (i.e. Relay Mining).
+// It starts and stops the RelayerProxy and provide the served relays observable to them miner.
+type relayMiner struct {
+	relayerProxy           RelayerProxy
+	miner                  Miner
+	relayerSessionsManager RelayerSessionsManager
+}
+
+// NewRelayMiner creates a new Relayer instance with the given dependencies.
+// It injects the dependencies into the Relayer instance and returns it.
+func NewRelayMiner(ctx context.Context, deps depinject.Config) (*relayMiner, error) {
+	rel := &relayMiner{}
+
+	if err := depinject.Inject(
+		deps,
+		&rel.relayerProxy,
+		&rel.miner,
+		&rel.relayerSessionsManager,
+	); err != nil {
+		return nil, err
+	}
+
+	// Set up relay pipeline
+	servedRelaysObs := rel.relayerProxy.ServedRelays()
+	minedRelaysObs := rel.miner.MinedRelays(ctx, servedRelaysObs)
+	rel.relayerSessionsManager.InsertRelays(minedRelaysObs)
+
+	return rel, nil
+}
+
+// Start provides the miner with the served relays observable and starts the relayer proxy.
+// This method is blocking while the relayer proxy is running and returns when Stop is called
+// or when the relayer proxy fails to start.
+func (rel *relayMiner) Start(ctx context.Context) error {
+	// relayerSessionsManager.Start does not block.
+	// Set up the session (proof/claim) lifecycle pipeline.
+	log.Println("INFO: Starting relayer sessions manager...")
+	rel.relayerSessionsManager.Start(ctx)
+
+	// Start the flow of relays by starting relayer proxy.
+	// This is a blocking call as it waits for the waitgroup to be done.
+	log.Println("INFO: Starting relayer proxy...")
+	if err := rel.relayerProxy.Start(ctx); err != nil {
```

Comment:
Is blocking inside of `Start` something we want to keep around long-term or just a current solution? Fine either way but just asking to understand where your mind is at.

Might be a `TODO_REVISIT` if you think it's worth it but not a blocker!

Reply:
I think we want to keep `Start` blocking as it's starting the relayer proxies. The passed context (from cmd) could close it anyway. And this is the top level function that is called from cmd. cc @bryanchriswhite

Reply:
Sgtm!